File: ASSIGNMNET6/python/Q.4.py
def precedence(op):
    """Return the precedence of the given operator."""
    if op == '+' or op == '-':
        return 1
    if op == '*' or op == '/':
        return 2
    return 0

def infix_to_postfix(expression):
    """Convert infix expression to postfix expression."""
    stack = []  # Stack to hold operators
    postfix = []  # List for output
    for char in expression:
        if char.isalnum():  # If the character is an operand (A-Z, a-z, 0-9)
            postfix.append(char)
        elif char == '(':  # If the character is '(', push it to stack
            stack.append(char)
        elif char == ')':  # If the character is ')'
            while stack and stack[-1] != '(':
                postfix.append(stack.pop())
            stack.pop()  # Pop '(' from stack
        else:  # If an operator is encountered
            while (stack and precedence(stack[-1]) >= precedence(char)):
                postfix.append(stack.pop())
            stack.append(char)

    # Pop all the operators from the stack
    while stack:
        postfix.append(stack.pop())
    
    return ''.join(postfix)

# Example usage
if __name__ == "__main__":
    expression = input("Enter an infix expression: ")
    postfix_expression = infix_to_postfix(expression)
    print("Postfix expression:", postfix_expression)

File: ASSIGNMNET6/c/Q.4.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#define MAX 100

// Stack structure definition
typedef struct {
    char items[MAX];
    int top;
} Stack;

// Function to initialize the stack
void initialize(Stack *s) {
    s->top = -1;
}

// Function to check if the stack is empty
int isEmpty(Stack *s) {
    return s->top == -1;
}

// Function to check if the stack is full
int isFull(Stack *s) {
    return s->top == MAX - 1;
}

// Function to push an element onto the stack
void push(Stack *s, char value) {
    if (isFull(s)) {
        printf("Stack overflow!\n");
    } else {
        s->items[++(s->top)] = value;
    }
}

// Function to pop an element from the stack
char pop(Stack *s) {
    if (isEmpty(s)) {
        printf("Stack underflow!\n");
        return '\0';
    } else {
        return s->items[(s->top)--];
    }
}

// Function to return the top element of the stack without popping
char peek(Stack *s) {
    if (isEmpty(s)) {
        return '\0';
    } else {
        return s->items[s->top];
    }
}

// Function to determine the precedence of operators
int precedence(char operator) {
    switch (operator) {
        case '^':
            return 3;
        case '*':
        case '/':
            return 2;
        case '+':
        case '-':
            return 1;
        default:
            return 0;
    }
}

// Function to check if a character is an operator
int isOperator(char ch) {
    return ch == '+' || ch == '-' || ch == '*' || ch == '/' || ch == '^';
}

// Function to convert an infix expression to a postfix expression
void infixToPostfix(char *infix, char *postfix) {
    Stack s;
    initialize(&s);
    int j = 0;  // Index for postfix expression

    for (int i = 0; i < strlen(infix); i++) {
        char ch = infix[i];

        // If the character is an operand, add it to the postfix expression
        if (isalnum(ch)) {
            postfix[j++] = ch;
        }
        // If the character is '(', push it to the stack
        else if (ch == '(') {
            push(&s, ch);
        }
        // If the character is ')', pop and output from the stack until '(' is found
        else if (ch == ')') {
            while (!isEmpty(&s) && peek(&s) != '(') {
                postfix[j++] = pop(&s);
            }
            pop(&s);  // Remove '(' from the stack
        }
        // If the character is an operator
        else if (isOperator(ch)) {
            while (!isEmpty(&s) && precedence(peek(&s)) >= precedence(ch)) {
                postfix[j++] = pop(&s);
            }
            push(&s, ch);
        }
    }

    // Pop all the remaining operators from the stack
    while (!isEmpty(&s)) {
        postfix[j++] = pop(&s);
    }

    postfix[j] = '\0';  // Null-terminate the postfix expression
}

// Main function to test the infix to postfix conversion
int main() {
    char infix[MAX], postfix[MAX];

    // Prompt user to enter an infix expression
    printf("Enter an infix expression: ");
    fgets(infix, sizeof(infix), stdin);

    // Remove newline character if present
    infix[strcspn(infix, "\n")] = '\0';

    infixToPostfix(infix, postfix);

    printf("Postfix expression: %s\n", postfix);

    return 0;
}
